File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

# 깊이 우선 탐색 (재귀로 구현)
def DFS(start):
    visited[start] = True
    print(start, end=" ")
    
    for i in graph[start]:
        if not visited[i]:
            DFS(i)

# 넓이 우선 탐색 (큐로 구현)
def BFS(start):
    queue = deque([start])
    visited[start] = True
    
    while queue:
        v = queue.popleft()
        print(v, end=" ")
        for i in graph[v]:
            if not visited[i]:
                visited[i] = True
                queue.append(i)

# 정점, 간선, 탐색을 시작할 정점 번호
N, M, V = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(N + 1)]

for _ in range(M):
    a, b = map(int, sys.stdin.readline().split())
    graph[a].append(b)
    graph[b].append(a)

# 정렬
for i in graph:
    i.sort()

# DFS
visited = [False] * (N+1)
DFS(V)

print()

# BFS
visited = [False] * (N+1)
BFS(V)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	
	static int N, M, V, arr[][];
	static boolean[] visited;
	static Queue<Integer> q = new LinkedList<>();
	static StringBuilder sb;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		sb = new StringBuilder();
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		V = Integer.parseInt(st.nextToken());
		arr = new int[N+1][N+1];
		
		for (int i=0; i<M; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			
			// 연결되어있다는 뜻
			arr[a][b] = 1;
			arr[b][a] = 1;
		}
		
		// DFS
		visited = new boolean[N+1];
		dfs(V);
		
		sb.append("\n");
		
		// BFS
		visited = new boolean[N+1];
		bfs(V);
		
		bw.write(sb.toString());
		bw.flush();
		bw.close();
		br.close();
	}
	
	public static void dfs(int start) {
		visited[start] = true;
		sb.append(start).append(" ");
		
		for (int i=0; i<=N; i++) {
			if (arr[start][i] == 1 && visited[i] == false) {
				dfs(i);
			}
		}
	}
	
	public static void bfs (int start) {
		visited[start] = true;
		q.add(start);
		
		while (!q.isEmpty()) {
			start = q.poll();
			sb.append(start).append(" ");
			for (int i=1; i<=N; i++) {
				if (arr[start][i] == 1 && visited[i] == false) {
					visited[i] = true;
					q.add(i);
				}
			}
		}
			
		
		
	}
}
